File: GoF Design Pattern/decorator.java
class Main {
  public static void main(String[] args) {
    Component component = new ConcreteComponent();
    component.operation();
    BuyGood phone = new BuyPhone();
    phone.displayPrice();
    
    Component component_withA = new ConcreteDecoratorA(component);
    component_withA.operation();
    BuyGood phone_withMicroSD = new BuyMicroSD(phone);
    phone_withMicroSD.displayPrice();
    
    Component component_withB = new ConcreteDecoratorB(component);
    component_withB.operation();
    BuyGood phone_withHeadset = new BuyHeadset(phone);
    phone_withHeadset.displayPrice();
    
    Component component_withA_andB = new ConcreteDecoratorB(component_withA);
    component_withA_andB.operation();
    BuyGood phone_withHeadset_andMicroSD = new BuyMicroSD(phone_withHeadset);
    phone_withHeadset_andMicroSD.displayPrice();
  }
}

abstract class Component {
  public abstract void operation();
}
abstract class BuyGood {
  public abstract int getTotalPrice();
  
  void displayPrice(){
    System.out.println("Total Price (IDR):");
    System.out.println(getTotalPrice());
  }
}

class ConcreteComponent extends Component {
  public void operation(){
    System.out.println("String from ConcreteComponent");
  }
}
class BuyPhone extends BuyGood {
  public int getTotalPrice(){
    return 2000000;
  }
}

abstract class Decorator extends Component {
  public abstract void operation();
}
abstract class BuyGoodDecorator extends BuyGood {
  public abstract int getTotalPrice();
}

class ConcreteDecoratorA extends Decorator {
  // aggregate Component
  Component component;
  
  public ConcreteDecoratorA(Component component){
    this.component = component;
  }
  
  public void operation(){
    component.operation();
    System.out.println("String from ConcreteDecoratorA");
  }
}
class BuyMicroSD extends BuyGoodDecorator {
  BuyGood good;
  
  public BuyMicroSD(BuyGood good){
    this.good = good;
  }
  
  public int getTotalPrice(){
    return 75000 + good.getTotalPrice();
  }
}

class ConcreteDecoratorB extends Decorator {
  // aggregate Component
  Component component;
  
  public ConcreteDecoratorB(Component component){
    this.component = component;
  }
  
  public void operation(){
    component.operation();
    System.out.println("String from ConcreteDecoratorB");
  }
}
class BuyHeadset extends BuyGoodDecorator {
  BuyGood good;
  
  public BuyHeadset(BuyGood good){
    this.good = good;
  }
  
  public int getTotalPrice(){
    return 30000 + good.getTotalPrice();
  }
}

File: GoF Design Pattern/decorator.js
function Component(){
  function operation(){
    console.log("String from ConcreteComponent")
  }
  
  function with_(...decoratorsArray){
    let behaviorArray = [operation]
    decoratorsArray.forEach(function(decorator){
      behaviorArray.push(decorator.addedBehavior)
    })
    return {
      operation: function(){
        // execute each behavior
        behaviorArray.forEach(function(behavior){behavior()})
      }
    }
  }
  
  return{with_, operation}
}
function Phone(){
  let price = 2000000

  function displayPrice(){
    let totalPrice = price
    console.log("Total Price (IDR):")
    console.log(totalPrice)
  }
  
  function with_(...phoneAccessories){
    let totalPrice = price
    phoneAccessories.forEach(
      (decorator) => {totalPrice = totalPrice + decorator.price}
    )
    
    return {displayPrice: function (){
        console.log("Total Price (IDR):")
        console.log(totalPrice)
      }
    }
  }
  
  return {with_, displayPrice}
}

let concreteDecoratorA = {
  addedBehavior: function(){console.log("String from ConcreteDecoratorA")}
}
let concreteDecoratorB = {
  addedBehavior: function(){console.log("String from ConcreteDecoratorB")}
}
let MicroSD = {price: 75000}
let Headset = {price: 30000}

let concreteComponent = Component()
concreteComponent.operation()
let phone = Phone()
phone.displayPrice()

concreteComponent.with_(concreteDecoratorA).operation()
phone.with_(MicroSD).displayPrice()

concreteComponent.with_(concreteDecoratorB).operation()
phone.with_(Headset).displayPrice()

concreteComponent.with_(concreteDecoratorA, concreteDecoratorB).operation()
phone.with_(Headset, MicroSD).displayPrice()